enterprise/cmd/repo-updater/authz/perms_syncer.go

```diff
@@ -0,0 +1,252 @@
+		UserID: userID,
+	})
+	if err != nil {
+		return errors.Wrap(err, "list external accounts")
+	}
+
+	// NOTE: The following logic is based on the assumption that we only support
+	// configuring one authz provider per code host type.
+	// See https://github.com/sourcegraph/sourcegraph/issues/8597 for details.
+	for _, acct := range accts {
+		fetcher := s.fetchers[acct.ServiceType]
+		if fetcher == nil {
+			// We have no authz provider configured for this external account.
+			continue
+		}
+
+		extIDs, err := fetcher.FetchUserPerms(ctx, acct)
+		if err != nil {
+			return errors.Wrap(err, "fetch user permissions")
+		}
+
+		// Get corresponding internal database IDs
+		repos, err := db.Repos.GetByExternalIDs(ctx, fetcher.ServiceType(), fetcher.ServiceID(), extIDs...)
+		if err != nil {
+			return errors.Wrap(err, "get repositories by external IDs")
+		}
+
+		// Save permissions to database
+		p := &authz.UserPermissions{
+			UserID:   userID,
+			Perm:     authz.Read, // Note: We currently only support read for repository permissions.
+			Type:     authz.PermRepos,
+			IDs:      roaring.NewBitmap(),
+			Provider: authz.ProviderType(fetcher.ServiceType()),
+		}
+		for i := range repos {
+			p.IDs.Add(uint32(repos[i].ID))
+		}
+
+		store := edb.NewPermsStore(s.db, s.clock)
+		err = store.SetUserPermissions(ctx, p)
+		if err != nil {
+			return errors.Wrap(err, "set user permissions")
+		}
+	}
+
+	return nil
+}
+
+// syncRepoPerms processes permissions syncing request in repository-centric way.
+// It discards requests that are made for non-private repositories based on the
+// value of "repo.private" column.
+func (s *PermsSyncer) syncRepoPerms(ctx context.Context, repoID api.RepoID) error {
+	repo, err := db.Repos.Get(ctx, repoID)
+	if err != nil {
+		return errors.Wrap(err, "get repositroy")
+	}
+
+	if !repo.Private {
+		return nil
+	}
+
+	fetcher := s.fetchers[repo.ExternalRepo.ServiceType]
+	if fetcher == nil {
+		// We have no authz provider configured for this repository.
+		return nil
+	}
+
+	// NOTE: The following logic is based on the assumption that we have accurate
+	// one-to-one username mapping between the internal database and the code host.
+	// See last paragraph of https://docs.sourcegraph.com/admin/auth#username-normalization
+	// for details.
+
+	usernames, err := fetcher.FetchRepoPerms(ctx, &repo.ExternalRepo)
+	if err != nil {
+		return errors.Wrap(err, "fetch repository permissions")
+	}
+
+	// Get corresponding internal database IDs
+	users, err := db.Users.GetByUsernames(ctx, usernames...)
+	if err != nil {
+		return errors.Wrap(err, "get users by usernames")
+	}
+
+	// Compute bind IDs for late-binding users
+	bindIDSet := make(map[string]struct{}, len(usernames))
+	for i := range usernames {
+		bindIDSet[usernames[i]] = struct{}{}
+	}
+	for i := range users {
+		delete(bindIDSet, users[i].Username)
+	}
+	pendingBindIDs := make([]string, 0, len(bindIDSet))
```

Comment:
```suggestion
	pendingBindUsernames := make([]string, 0, len(bindIDSet))
```

Unless I'm missing something, these are usernames, right?